### PR TITLE
gh-176: LCDM, wCDM and waw0CDM linear emulator module

### LINKED ISSUE #176: CosmoPower new emulators beyond LCDM
## 📝 Issue Summary
I wish  to add a few matter power spectrum emulators to the repository. 

### 🎯 Goals / Acceptance Criteria
- [ ] **Goal 1**:  Adding the cosmopower_cosmology module
- [ ] **Goal 2**: Expanding the number of emulators (to be added at a later date)

### 🔄 Steps to Reproduce (if applicable)
1.Merge the emulator module

### ✅ Expected Outcome
Making the matter power spectrum emulators available to the Euclid CLOE organisation

### 📌 Additional Notes
Currently, there are LCDM and w0wa cosmology emulators available (both linear and nonlinear), with different neutrino models (either no massive neutrinos, 1 massive neutrino and 2 massless, and 3 degenerate massive neutrinos).

### PR DESCRIPTION
**Pull request by ivansladoljev**
### ✅ Summary
Adding module for w0wa, wcdm and LCDM emulators into cloelib. This modules contains several pre trained neural emulators that calculate the linear and nonlinear matter power spectra [(both full matter and cb [cold dark matter+baryons)] for different cosmologies. 

These emulators are made high accuracy predictions from CAMB, using very large parameter prior ranges. Special care was taken to model the neutrino physics correctly.

### 🔄 Changes
``cloelib/cosmology/cosmopower_cosmology.py `` module for cosmological emulators. It contains several trained emulators for the matter power spectrum made with Cosmopower package.
 Added ``test_emulators.py`` script in the ``cloelib/test folder``  for basic unit tests.

### 🛠 How to Test
An example of an emulator call can be seen here `` from cloelib.cosmology.cosmopower_cosmology import
        CosmoPowerw0waCDMPerturbations as w0waCDM, emulator= w0waCDM.Linear(background=camb_instance, redshifts=z); pk = emulator.matter_power_spectrum(0, k_emu)[0, :]``

Depending on the ``N_mnu`` parameter in the background perturbation, the ``Linear`` class will assign the correct pre-trained emulator for that model (0 for massless neutrinos, 1 for single massive neutrino and 3 for 3 degenerate neutrinos). If the user wishes to have the CB (cold dark matter + baryon) matter power spectrum, they should choose ``LinearCB`` instead.

I have added a unit test script (can be updated if needed) located in the test folder. When merged, I can update the playground notebook as well.

### 📝 Documentation
- [ ] This PR does not require documentation changes

### 🏗 Related Issues

Resolves #176 

### 📌 Additional Notes

lcdm-plin

The picture shows the absolute difference between the emulator predictions and 50000 test spectra, produced with high accuracy settings from CAMB. The model in question is LCDM linear power spectrum.

---

### ✅ PR Checklist for Developers
- [x] I have titled this PR before merging as "gh-#:", where "#" represents the task it closes
- [x] My code follows the repository's coding style
- [x] I have tested my changes locally
- [x] No new warnings or errors introduced
- [x] I have updated documentation (if applicable)
- [x] My changes do not introduce breaking changes
- [x] I have added tests (if applicable)
- [x] I have consistently updated the GitHub information for the project, including milestones, task types, and other relevant details.

### ✅ PR Checklist for Reviewers
- [ ] The next PR targets the correct branch
- [ ] CI tests have run and passed for the latest commit on the source branch
- [ ] Check that the code can still be installed if new packages are imported
- [ ] If necessary, the notebooks in the `playground` will be updated in a corresponding follow-up PR
- [ ] Coverage percentage is retained or increased
- [ ] Quality of new/changed code is acceptable
- [ ] Quality of new/changed unit tests is acceptable
- [ ] No data files have been included in the commits
- [ ] Implementation follows the agreed task description point by point
- [ ] Check that there are no `No newline at the end of file` warnings
- [ ] Check that any added folder/file has been added to the `README.md` file
- [ ] Check that the implementation follows the contributing guidelines and style choices
- [ ] Check that the documentation has been updated accordantly
- [ ] Check that the corresponding branch has been deleted after merging. If not, delete it

### CODE REVIEW

File: cloelib/cosmology/cosmopower_cosmology.py [hunk 1]
```diff
@@ -0,0 +1,954 @@
...
+from cloelib.auxiliary.extrapolator import extend_spectra
+
+import numpy as np
+import warnings
+from scipy import interpolate
+import os
+import urllib.request
+
+os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
+import tensorflow as tf
```
Comment (gcanasherrera) [inline cloelib/cosmology/cosmopower_cosmology.py]: Hi @ivansladoljev , can you comment which tensorflow version do we still need?

File: cloelib/cosmology/cosmopower_cosmology.py [hunk 2]
```diff
@@ -0,0 +1,954 @@
...
+        file_path = file_path
+
+    return file_path
+
+
+zenodo_path = "https://zenodo.org/records/17570978/files"
+k_modes_path = emulator_data("k-modes.txt", zenodo_path)
+
+
+class w0waCDM_Linear:
```
Comment (gcanasherrera) [inline cloelib/cosmology/cosmopower_cosmology.py]: We need to think a bit more about the names, so they are compatible with other Perturbations. What about:

`CosmoPowerw0waCDMPerturbations` ? It's too long, but it's similar to `CAMBPerturbations` etc. Ideas?

File: cloelib/cosmology/cosmopower_cosmology.py [hunk 3]
```diff
@@ -0,0 +1,954 @@
...
+        Raises
+        ------
+        AssertionError
+            If the geometry is not flat (Omega_k0 != 0).
+        ValueError
+            If any parameter lies outside the bounds supported by the emulator.
+        """
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", category=UserWarning)
+            import cosmopower as cp
```
Comment (gcanasherrera) [inline cloelib/cosmology/cosmopower_cosmology.py]: I'll try to import it when this file is imported first, not when you initialize, similarly to other external packages

**Review by gcanasherrera — CHANGES_REQUESTED** [review]
Just first initial comments

File: cloelib/cosmology/cosmopower_cosmology.py (hunk 1 above)
Comment (ivansladoljev) [reply cloelib/cosmology/cosmopower_cosmology.py]: I am using 2.13.1 at the moment, but there is a plan to fix the problems caused with tensorflow dependancies in the near future.

File: cloelib/cosmology/cosmopower_cosmology.py (hunk 3 above)
Comment (ivansladoljev) [reply cloelib/cosmology/cosmopower_cosmology.py]: Thank you, I will change it immediately.

File: cloelib/cosmology/cosmopower_cosmology.py (hunk 2 above)
Comment (ivansladoljev) [reply cloelib/cosmology/cosmopower_cosmology.py]: I agree with you fully. I will change the names of  all my classes. The only problem with this naming convention I might foresee is when we start adding different neutrino modelling. Then my classes would be called ``CosmoPowerw0waCDM1massPerturbations`` or ``CosmoPowerw0waCDM3degenPerturbations``, not to mention nonlinear and Pcb variants ... 
Might be good to think of a shorter way to write it.

Comment (ivansladoljev) [reply cloelib/cosmology/cosmopower_cosmology.py]: Since there are 30+ different emulators I made, I might need to think of a smarter way to group them, so people don't get lost amongst it all.

Comment (gcanasherrera) [reply cloelib/cosmology/cosmopower_cosmology.py]: Indeed, this is something that worries me a bit. Have you thought about how to interface this directly with the public with cosmopower?

File: cloelib/cosmology/cosmopower_cosmology.py (hunk 1 above)
Comment (gcanasherrera) [reply cloelib/cosmology/cosmopower_cosmology.py]: Thanks!

File: cloelib/cosmology/cosmopower_cosmology.py (hunk 2 above)
Comment (ivansladoljev) [reply cloelib/cosmology/cosmopower_cosmology.py]: @gcanasherrera well currently no, but that might be something we do in the future. For now the quickest fix would be to introduce subclasses, so for example the user would import a specific cosmology in mind ``from cloelib.cosmology.cosmopower_cosmology import CosmoPowerLCDMPerturbations`` and then type ``linear_emu =  CosmoPowerLCDMPerturbations.Linear(background=camb_instance, redshifts=z)`` or  `` linear_emu_with_massive =  CosmoPowerLCDMPerturbations.1mass_Linear(background=camb_instance, redshifts=z)``. I don't know if that is allowed under our naming conventions ?

Also, this problem touches on something @arthurmloureiro talked about - to have a single framework for calling neutrinos. For example, having a yaml file with the path to the code, priors, bounds, etc - and having a simple script that reads he file and outputs the spectra. That might be a pain to implement, but could be worthwhile - as there are currently dozens of different emulators waiting to be reviewed, each with its own different script....,

**Comment by ivansladoljev** [comment]
Hi @gcanasherrera, so yesterday I added the new sub-classes, for example ``CosmoPowerLCDPerturbations.Linear`` and today I added the 1 massive neutrino ones as well (for example ``CosmoPowerLCDPerturbations.Linear_1mass``). Hope this is a better system, than having one class per emulator. 

If this is a little bit overwhelming to review at once, I can always revert back to the old git commit.

Looking forward to your review, and also I really think we should have a broader discussion on how to structure the every emulator in Euclid a similar way, so people have easier time using it.

**Comment by ivansladoljev** [comment]
Hi @gcanasherrera , I found a new way to sort through the emulator classes. Instead of having one class per emulator (for 40+ emulators), I compacted that into one class per cosmology. So if somebody wants for example the linear spectra they call a cosmology `` from cloelib.cosmology.cosmopower_cosmology import CosmoPowerLCDMPerturbations as LCDM`` and call an emulator as a subclass of that ``LCDM.Linear, LCDM.LinearCB, LCDM.NonLinear, etc`. And depending what ``N_mnu`` parameter they put in the background perturbations, the module will automatically give them the correct emulator.

**Review by gcanasherrera — CHANGES_REQUESTED** [review]
@ivansladoljev can you change the name of the tests? Not just emulators, but call it cosmopower somehow

**Comment by ivansladoljev** [comment]
> @ivansladoljev can you change the name of the tests? Not just emulators, but call it cosmopower somehow

Of course, no problem. I also wanted to ask whether there is a default set of tests that need to be implemented, or is every module going to have its own set of tests ?

**Comment by gcanasherrera** [comment]
The idea is that, usually, we have tests per module.

**Comment by ivansladoljev** [comment]
> The idea is that, usually, we have tests per module.

Changed. I also removed the old cosmopower module so now there is only cosmpower_jax ones.

**Comment by gcanasherrera** [comment]
ok @ivansladoljev thanks a lot! I'm going to merge, and then, do you mind adding some docs? I'll tell you more in Slack :)

**Comment by gcanasherrera** [comment]
@all-contributors please add @ivansladoljev for code, ideas

**Comment by ivansladoljev** [comment]
> ok @ivansladoljev thanks a lot! I'm going to merge, and then, do you mind adding some docs? I'll tell you more in Slack :)

Of course, no problem :)
